packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1210 @@
+    }
+    return id;
+  }
+
+  /**
+   * Adds an edge to the graph.
+   *
+   * Returns `true` if the edge was added,
+   * or `false` if the edge already exists.
+   */
+  addEdge(
+    from: NodeId,
+    to: NodeId,
+    type: TEdgeType | NullEdgeType = 1,
+  ): boolean {
+    assert(type > 0, `Unsupported edge type ${0}`);
+
+    let hash = this.#edges.hash(from, to, type);
+    let edge = this.#edges.addressOf(hash, from, to, type);
+
+    // The edge is already in the graph; do nothing.
+    if (edge !== null) return false;
+
+    let capacity = this.#edges.capacity;
+    // We add 1 to account for the edge we are adding.
+    let count = this.#edges.count + 1;
+    // Since the space occupied by deleted edges isn't reclaimed,
+    // we include them in our count to avoid overflowing the `edges` array.
+    let deletes = this.#edges.deletes;
+    let total = count + deletes;
+    // If we have enough space to keep adding edges, we can
+    // put off reclaiming the deleted space until the next resize.
+    if (this.#edges.getLoad(total) > LOAD_FACTOR) {
+      if (this.#edges.getLoad(deletes) > UNLOAD_FACTOR) {
+        // If we have a significant number of deletes, we compute our new
+        // capacity based on the current count, even though we decided to
+        // resize based on the sum total of count and deletes.
+        // In this case, resizing is more like a compaction.
+        this.resizeEdges(
+          getNextEdgeCapacity(capacity, count, this.#edges.getLoad(count)),
+        );
+      } else {
+        this.resizeEdges(
+          getNextEdgeCapacity(capacity, total, this.#edges.getLoad(total)),
+        );
+      }
+      // We must rehash because the capacity has changed.
+      hash = this.#edges.hash(from, to, type);
+    }
+
+    let toNode = this.#nodes.addressOf(to, type);
+    let fromNode = this.#nodes.addressOf(from, type);
+    if (toNode === null || fromNode === null) {
+      // If we're in danger of overflowing the `nodes` array, resize it.
+      if (this.#nodes.load >= LOAD_FACTOR) {
+        this.resizeNodes(increaseNodeCapacity(this.#nodes.capacity));
+        // We need to update our indices since the `nodes` array has changed.
+        toNode = this.#nodes.addressOf(to, type);
+        fromNode = this.#nodes.addressOf(from, type);
+      }
+    }
+    if (toNode === null) toNode = this.#nodes.add(to, type);
+    if (fromNode === null) fromNode = this.#nodes.add(from, type);
+
+    // Add our new edge to its hash bucket.
+    edge = this.#edges.add(hash, from, to, type);
+
+    // Link this edge to the node's list of incoming edges.
+    let prevIn = this.#nodes.linkIn(toNode, edge);
+    if (prevIn !== null) this.#edges.linkIn(prevIn, edge);
+
+    // Link this edge to the node's list of outgoing edges.
+    let prevOut = this.#nodes.linkOut(fromNode, edge);
+    if (prevOut !== null) this.#edges.linkOut(prevOut, edge);
+
+    return true;
+  }
+
+  *getAllEdges(): Iterator<{|
```

Comment:
This is the now the only scenario where we generate values for iteration. All other traversals are now using arrays.